# ETL Pipeline Preparation
Follow the instructions below to help you create your ETL pipeline.
### 1. Import libraries and load datasets.
- Import Python libraries
- Load `messages.csv` into a dataframe and inspect the first few lines.
- Load `categories.csv` into a dataframe and inspect the first few lines.

In [1]:
# import libraries
import pandas as pd
import numpy as np
from collections import defaultdict
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_squared_error
import matplotlib.pyplot as plt

In [3]:
# load messages dataset
messages = pd.read_csv('messages.csv', index_col=0)
print(messages.shape)
messages.head()

(26248, 3)


message  \
id                                                      
2   Weather update - a cold front from Cuba that c...   
7             Is the Hurricane over or is it not over   
8                     Looking for someone but no name   
9   UN reports Leogane 80-90 destroyed. Only Hospi...   
12  says: west side of Haiti, rest of the country ...   

                                             original   genre  
id                                                             
2   Un front froid se retrouve sur Cuba ce matin. ...  direct  
7                  Cyclone nan fini osinon li pa fini  direct  
8   Patnm, di Maryani relem pou li banm nouvel li ...  direct  
9   UN reports Leogane 80-90 destroyed. Only Hospi...  direct  
12  facade ouest d Haiti et le reste du pays aujou...  direct

In [4]:
# load categories dataset
categories = pd.read_csv('categories.csv', index_col=0)
print(categories.shape)
categories.head()

(26248, 1)


categories
id                                                   
2   related-1;request-0;offer-0;aid_related-0;medi...
7   related-1;request-0;offer-0;aid_related-1;medi...
8   related-1;request-0;offer-0;aid_related-0;medi...
9   related-1;request-1;offer-0;aid_related-1;medi...
12  related-1;request-0;offer-0;aid_related-0;medi...

### 2. Merge datasets.
- Merge the messages and categories datasets using the common id
- Assign this combined dataset to `df`, which will be cleaned in the following steps

In [5]:
# merge datasets
#df = messages.merge(categories, left_on='id', right_on='id', how='outer')
df = messages.merge(categories, how = 'left', on = ['id'])
df.head()

message  \
id                                                      
2   Weather update - a cold front from Cuba that c...   
7             Is the Hurricane over or is it not over   
8                     Looking for someone but no name   
9   UN reports Leogane 80-90 destroyed. Only Hospi...   
12  says: west side of Haiti, rest of the country ...   

                                             original   genre  \
id                                                              
2   Un front froid se retrouve sur Cuba ce matin. ...  direct   
7                  Cyclone nan fini osinon li pa fini  direct   
8   Patnm, di Maryani relem pou li banm nouvel li ...  direct   
9   UN reports Leogane 80-90 destroyed. Only Hospi...  direct   
12  facade ouest d Haiti et le reste du pays aujou...  direct   

                                           categories  
id                                                     
2   related-1;request-0;offer-0;aid_related-0;medi...  
7   related-1;request-0;offer-0;aid_related-1;medi...  
8   related-1;request-0;offer-0;aid_related-0;medi...  
9   related-1;request-1;offer-0;aid_related-1;medi...  
12  related-1;request-0;offer-0;aid_related-0;medi...

### 3. Split `categories` into separate category columns.
- Split the values in the `categories` column on the `;` character so that each value becomes a separate column. You'll find [this method](https://pandas.pydata.org/pandas-docs/version/0.23/generated/pandas.Series.str.split.html) very helpful! Make sure to set `expand=True`.
- Use the first row of categories dataframe to create column names for the categories data.
- Rename columns of `categories` with new column names.

In [6]:
# create a dataframe of the 36 individual category columns
categories = df['categories'].str.split(';', expand=True)
categories.head()

0          1        2              3               4   \
id                                                                 
2   related-1  request-0  offer-0  aid_related-0  medical_help-0   
7   related-1  request-0  offer-0  aid_related-1  medical_help-0   
8   related-1  request-0  offer-0  aid_related-0  medical_help-0   
9   related-1  request-1  offer-0  aid_related-1  medical_help-0   
12  related-1  request-0  offer-0  aid_related-0  medical_help-0   

                    5                    6           7           8   \
id                                                                    
2   medical_products-0  search_and_rescue-0  security-0  military-0   
7   medical_products-0  search_and_rescue-0  security-0  military-0   
8   medical_products-0  search_and_rescue-0  security-0  military-0   
9   medical_products-1  search_and_rescue-0  security-0  military-0   
12  medical_products-0  search_and_rescue-0  security-0  military-0   

               9   ...             26                      27  \
id                 ...                                          
2   child_alone-0  ...  aid_centers-0  other_infrastructure-0   
7   child_alone-0  ...  aid_centers-0  other_infrastructure-0   
8   child_alone-0  ...  aid_centers-0  other_infrastructure-0   
9   child_alone-0  ...  aid_centers-0  other_infrastructure-0   
12  child_alone-0  ...  aid_centers-0  other_infrastructure-0   

                   28        29       30      31            32      33  \
id                                                                       
2   weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   
7   weather_related-1  floods-0  storm-1  fire-0  earthquake-0  cold-0   
8   weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   
9   weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   
12  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   

                 34               35  
id                                    
2   other_weather-0  direct_report-0  
7   other_weather-0  direct_report-0  
8   other_weather-0  direct_report-0  
9   other_weather-0  direct_report-0  
12  other_weather-0  direct_report-0  

[5 rows x 36 columns]

In [8]:
# select the first row of the categories dataframe
row = categories.iloc[0, ]
category_colnames = row.apply(lambda x: x.split('-')[0]).tolist()
print(category_colnames)

['related', 'request', 'offer', 'aid_related', 'medical_help', 'medical_products', 'search_and_rescue', 'security', 'military', 'child_alone', 'water', 'food', 'shelter', 'clothing', 'money', 'missing_people', 'refugees', 'death', 'other_aid', 'infrastructure_related', 'transport', 'buildings', 'electricity', 'tools', 'hospitals', 'shops', 'aid_centers', 'other_infrastructure', 'weather_related', 'floods', 'storm', 'fire', 'earthquake', 'cold', 'other_weather', 'direct_report']


In [9]:
# rename the columns of `categories`
categories.columns = category_colnames
categories.head()

related    request    offer    aid_related    medical_help  \
id                                                                 
2   related-1  request-0  offer-0  aid_related-0  medical_help-0   
7   related-1  request-0  offer-0  aid_related-1  medical_help-0   
8   related-1  request-0  offer-0  aid_related-0  medical_help-0   
9   related-1  request-1  offer-0  aid_related-1  medical_help-0   
12  related-1  request-0  offer-0  aid_related-0  medical_help-0   

      medical_products    search_and_rescue    security    military  \
id                                                                    
2   medical_products-0  search_and_rescue-0  security-0  military-0   
7   medical_products-0  search_and_rescue-0  security-0  military-0   
8   medical_products-0  search_and_rescue-0  security-0  military-0   
9   medical_products-1  search_and_rescue-0  security-0  military-0   
12  medical_products-0  search_and_rescue-0  security-0  military-0   

      child_alone  ...    aid_centers    other_infrastructure  \
id                 ...                                          
2   child_alone-0  ...  aid_centers-0  other_infrastructure-0   
7   child_alone-0  ...  aid_centers-0  other_infrastructure-0   
8   child_alone-0  ...  aid_centers-0  other_infrastructure-0   
9   child_alone-0  ...  aid_centers-0  other_infrastructure-0   
12  child_alone-0  ...  aid_centers-0  other_infrastructure-0   

      weather_related    floods    storm    fire    earthquake    cold  \
id                                                                       
2   weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   
7   weather_related-1  floods-0  storm-1  fire-0  earthquake-0  cold-0   
8   weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   
9   weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   
12  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   

      other_weather    direct_report  
id                                    
2   other_weather-0  direct_report-0  
7   other_weather-0  direct_report-0  
8   other_weather-0  direct_report-0  
9   other_weather-0  direct_report-0  
12  other_weather-0  direct_report-0  

[5 rows x 36 columns]

### 4. Convert category values to just numbers 0 or 1.
- Iterate through the category columns in df to keep only the last character of each string (the 1 or 0). For example, `related-0` becomes `0`, `related-1` becomes `1`. Convert the string to a numeric value.
- You can perform [normal string actions on Pandas Series](https://pandas.pydata.org/pandas-docs/stable/text.html#indexing-with-str), like indexing, by including `.str` after the Series. You may need to first convert the Series to be of type string, which you can do with `astype(str)`.

In [11]:
for column in categories:
    # set each value to be the last character of the string
    categories[column] = categories[column].apply(lambda x: x[-1])
    # convert column from string to numeric
    #categories[column] = categories[column].astype(int)
    categories[column] = pd.to_numeric(categories[column])
print(categories.dtypes)
categories.head()

related                   int64
request                   int64
offer                     int64
aid_related               int64
medical_help              int64
medical_products          int64
search_and_rescue         int64
security                  int64
military                  int64
child_alone               int64
water                     int64
food                      int64
shelter                   int64
clothing                  int64
money                     int64
missing_people            int64
refugees                  int64
death                     int64
other_aid                 int64
infrastructure_related    int64
transport                 int64
buildings                 int64
electricity               int64
tools                     int64
hospitals                 int64
shops                     int64
aid_centers               int64
other_infrastructure      int64
weather_related           int64
floods                    int64
storm                     int64
fire    

related  request  offer  aid_related  medical_help  medical_products  \
id                                                                         
2         1        0      0            0             0                 0   
7         1        0      0            1             0                 0   
8         1        0      0            0             0                 0   
9         1        1      0            1             0                 1   
12        1        0      0            0             0                 0   

    search_and_rescue  security  military  child_alone  ...  aid_centers  \
id                                                      ...                
2                   0         0         0            0  ...            0   
7                   0         0         0            0  ...            0   
8                   0         0         0            0  ...            0   
9                   0         0         0            0  ...            0   
12                  0         0         0            0  ...            0   

    other_infrastructure  weather_related  floods  storm  fire  earthquake  \
id                                                                           
2                      0                0       0      0     0           0   
7                      0                1       0      1     0           0   
8                      0                0       0      0     0           0   
9                      0                0       0      0     0           0   
12                     0                0       0      0     0           0   

    cold  other_weather  direct_report  
id                                      
2      0              0              0  
7      0              0              0  
8      0              0              0  
9      0              0              0  
12     0              0              0  

[5 rows x 36 columns]

### 5.Data Exploration

In [12]:
# Confirm values in each column of the categories dataframe are, in fact, only 0's and 1's.
for column in categories:
    print(column)
    print(categories[column].value_counts())

related
1    20042
0     6140
2      204
Name: related, dtype: int64
request
0    21873
1     4513
Name: request, dtype: int64
offer
0    26265
1      121
Name: offer, dtype: int64
aid_related
0    15432
1    10954
Name: aid_related, dtype: int64
medical_help
0    24287
1     2099
Name: medical_help, dtype: int64
medical_products
0    25067
1     1319
Name: medical_products, dtype: int64
search_and_rescue
0    25661
1      725
Name: search_and_rescue, dtype: int64
security
0    25915
1      471
Name: security, dtype: int64
military
0    25523
1      863
Name: military, dtype: int64
child_alone
0    26386
Name: child_alone, dtype: int64
water
0    24702
1     1684
Name: water, dtype: int64
food
0    23430
1     2956
Name: food, dtype: int64
shelter
0    24044
1     2342
Name: shelter, dtype: int64
clothing
0    25976
1      410
Name: clothing, dtype: int64
money
0    25780
1      606
Name: money, dtype: int64
missing_people
0    26085
1      301
Name: missing_people, dtype: int64
refuge

In [13]:
# Count number of entries where related == 2
categories['related'].value_counts()

1    20042
0     6140
2      204
Name: related, dtype: int64

> There are 204 cases where the message has a "related" value of 2. We shall explore what these messages look like, and determine what to do with them, in the next section.

> All values in the child_alone column are equal to 0. As a result, based on this data, the best model we will be able to devise to predict whether or not the message indicates that a child is alone will just predict that it is not 100% of the time. As such a model is pointless, we will drop this column from the categories dataframe.

In [14]:
# Drop child_alone from categories dataframe.
categories.drop('child_alone', axis = 1, inplace = True)

categories.head()

related  request  offer  aid_related  medical_help  medical_products  \
id                                                                         
2         1        0      0            0             0                 0   
7         1        0      0            1             0                 0   
8         1        0      0            0             0                 0   
9         1        1      0            1             0                 1   
12        1        0      0            0             0                 0   

    search_and_rescue  security  military  water  ...  aid_centers  \
id                                                ...                
2                   0         0         0      0  ...            0   
7                   0         0         0      0  ...            0   
8                   0         0         0      0  ...            0   
9                   0         0         0      0  ...            0   
12                  0         0         0      0  ...            0   

    other_infrastructure  weather_related  floods  storm  fire  earthquake  \
id                                                                           
2                      0                0       0      0     0           0   
7                      0                1       0      1     0           0   
8                      0                0       0      0     0           0   
9                      0                0       0      0     0           0   
12                     0                0       0      0     0           0   

    cold  other_weather  direct_report  
id                                      
2      0              0              0  
7      0              0              0  
8      0              0              0  
9      0              0              0  
12     0              0              0  

[5 rows x 35 columns]

### 6. Replace `categories` column in `df` with new category columns.
- Drop the categories column from the df dataframe since it is no longer needed.
- Concatenate df and categories data frames.

In [15]:
# drop the original categories column from `df`
df.drop('categories', axis = 1, inplace = True)
df.head()

message  \
id                                                      
2   Weather update - a cold front from Cuba that c...   
7             Is the Hurricane over or is it not over   
8                     Looking for someone but no name   
9   UN reports Leogane 80-90 destroyed. Only Hospi...   
12  says: west side of Haiti, rest of the country ...   

                                             original   genre  
id                                                             
2   Un front froid se retrouve sur Cuba ce matin. ...  direct  
7                  Cyclone nan fini osinon li pa fini  direct  
8   Patnm, di Maryani relem pou li banm nouvel li ...  direct  
9   UN reports Leogane 80-90 destroyed. Only Hospi...  direct  
12  facade ouest d Haiti et le reste du pays aujou...  direct

In [17]:
# concatenate the original dataframe with the new `categories` dataframe
df = pd.concat([df, categories], axis = 1)
df.head()

message  \
id                                                      
2   Weather update - a cold front from Cuba that c...   
7             Is the Hurricane over or is it not over   
8                     Looking for someone but no name   
9   UN reports Leogane 80-90 destroyed. Only Hospi...   
12  says: west side of Haiti, rest of the country ...   

                                             original   genre  related  \
id                                                                       
2   Un front froid se retrouve sur Cuba ce matin. ...  direct        1   
7                  Cyclone nan fini osinon li pa fini  direct        1   
8   Patnm, di Maryani relem pou li banm nouvel li ...  direct        1   
9   UN reports Leogane 80-90 destroyed. Only Hospi...  direct        1   
12  facade ouest d Haiti et le reste du pays aujou...  direct        1   

    request  offer  aid_related  medical_help  medical_products  \
id                                                                
2         0      0            0             0                 0   
7         0      0            1             0                 0   
8         0      0            0             0                 0   
9         1      0            1             0                 1   
12        0      0            0             0                 0   

    search_and_rescue  ...  aid_centers  other_infrastructure  \
id                     ...                                      
2                   0  ...            0                     0   
7                   0  ...            0                     0   
8                   0  ...            0                     0   
9                   0  ...            0                     0   
12                  0  ...            0                     0   

    weather_related  floods  storm  fire  earthquake  cold  other_weather  \
id                                                                          
2                 0       0      0     0           0     0              0   
7                 1       0      1     0           0     0              0   
8                 0       0      0     0           0     0              0   
9                 0       0      0     0           0     0              0   
12                0       0      0     0           0     0              0   

    direct_report  
id                 
2               0  
7               0  
8               0  
9               0  
12              0  

[5 rows x 38 columns]

### 7. Remove duplicates.
- Check how many duplicates are in this dataset.
- Drop the duplicates.
- Confirm duplicates were removed.

In [19]:
# check number of duplicates
print(df.duplicated().sum())
print(df.shape)

171
(26386, 38)


In [20]:
# drop duplicates
df = df.drop_duplicates()
df.shape

(26215, 38)

In [21]:
# check number of duplicates
print(df.duplicated().sum())

0


### 7. Explore messages with a "related" value of 2.¶


In [22]:
df['message'][df['related'] == 2]

id
146      Dans la zone de Saint Etienne la route de Jacm...
263      . .. i with limited means. Certain patients co...
373      The internet caf Net@le that's by the Dal road...
565      Bonsoir, on est a bon repos aprs la compagnie ...
700      URGENT CRECHE ORPHANAGE KAY TOUT TIMOUN CROIX ...
                               ...                        
23662         transferred to a sanitary landfill site by a
23861    Families also have solar lamps which can be re...
25874    Read the [full blog post](http://www.odi.org.u...
27071    Actualmente e independientemente de la ayuda d...
29126    Mali's former President Amadou Toumani Touré -...
Name: message, Length: 188, dtype: object

> Many of the messages with a related value of 2 appear to be in languages other than English. Based on this, we infer that a related value of 2 corresponds to uncertainty as to whether or not the message is related. As a result, these records are not really useful when it comes to building an ML model, so we will drop these records from the dataset.

In [23]:
# Remove rows with a related value of 2 from the dataset
df = df[df['related'] != 2]

### 8. Save the clean dataset into an sqlite database.
You can do this with pandas [`to_sql` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.to_sql.html) combined with the SQLAlchemy library. Remember to import SQLAlchemy's `create_engine` in the first cell of this notebook to use it below.

In [25]:
from sqlalchemy import create_engine
engine = create_engine('sqlite:///DisasterMsgTable.db')
df.to_sql('DisasterMsgTable', engine, index=False)

### 8. Use this notebook to complete `etl_pipeline.py`
Use the template file attached in the Resources folder to write a script that runs the steps above to create a database based on new datasets specified by the user. Alternatively, you can complete `etl_pipeline.py` in the classroom on the `Project Workspace IDE` coming later.